fdb-record-layer-core/src/test/java/com/apple/foundationdb/record/cursors/ConcatCursorTest.java

```diff
@@ -0,0 +1,438 @@
+        TriFunction<FDBRecordContext, ScanProperties, byte[], RecordCursor<KeyValue>> fcc2 = (c, s, b) -> KeyValueCursor.Builder.withSubspace(subspace).setScanProperties(s).setContext(c)
+                .setRange(TupleRange.ALL).setContinuation(b).build();
+
+        Integer firstCursorSize = fcc1.apply(context, ScanProperties.FORWARD_SCAN, null).getCount().join();
+        Integer secondCursorSize = fcc1.apply(context, ScanProperties.FORWARD_SCAN, null).getCount().join();
+        Integer concatCursorSize = firstCursorSize + secondCursorSize;
+        Integer expectedResultSize;
+        Integer rowLimit = scanProperties.getExecuteProperties().getReturnedRowLimit();
+        if (rowLimit > 0 && rowLimit <= concatCursorSize) {
+            expectedResultSize = Integer.min(rowLimit, concatCursorSize);
+        } else {
+            expectedResultSize = concatCursorSize; // no row limit
+        }
+
+        //concat a couple of kv cursors and check that we get expected number of rows
+        cc = new ConcatCursor<>(context, scanProperties, fcc1, fcc2, null);
+        Assertions.assertEquals(expectedResultSize, (Integer)cc.asList().join().size());
+
+        //now run the same tests breaking up the concat cursor runs and use continuation to complete
+        //stop and restart the cursor at each each possible location in the sequence
+        Integer stop;
+        TestResult testResult1;
+        TestResult testResult2;
+        Integer totalRowsScanned;
+        for (stop = 1; stop < concatCursorSize; stop++) {
+            cc = new ConcatCursor<>(context, scanProperties, fcc1, fcc2, null);
+            testResult1 = synchronousIterationAndCompare(null, cc, stop);
+            cc = new ConcatCursor<>(context, scanProperties, fcc1, fcc2, testResult1.getContinuation());
+            testResult2 = synchronousIterationAndCompare(null, cc, expectedResultSize);
+            totalRowsScanned = testResult1.getCount() + testResult2.getCount();
+            Assertions.assertEquals(totalRowsScanned, (Integer)(rowLimit == 0 ? concatCursorSize : Integer.min((Integer.min(stop, rowLimit) + rowLimit), concatCursorSize)));  //each invocation can only get as far as rowLimit
+
+            cc = new ConcatCursor<>(context, scanProperties, fcc1, fcc2, null);
+            testResult1 = asynchronousIterationAndCompare(null, cc, stop);
+            cc = new ConcatCursor<>(context, scanProperties, fcc1, fcc2, testResult1.getContinuation());
+            testResult2 = asynchronousIterationAndCompare(null, cc, expectedResultSize);
+            totalRowsScanned = testResult1.getCount() + testResult2.getCount();
+            Assertions.assertEquals(totalRowsScanned, (Integer)(rowLimit == 0 ? concatCursorSize : Integer.min((Integer.min(stop, rowLimit) + rowLimit), concatCursorSize)));  //each invocation can only get as far as rowLimit
+        }
+    }
+
+    public void concatTimeLimitTest(ScanProperties scanProperties) {
```

Comment:
I'm not sure how this is actually exercising a time limit--meaning, how do you even know that the time limit is being hit at all?   The usual trick that I do is to have a given cursor add an artificial delay like:

    KeyValueCursor.Builder(...).build().map(v -> { sleep(1L); return v; }

so each fetch will take about a millisecond.  Then set a time limit of, say, 5ms.  Then sit in a loop over all of the continuations until you have consumed the full list and simply make sure that you have received at least one or two continuations.

Reply:
done